File: frontend/src/types/spatial.ts
import type { FeatureCollection, Point } from 'geojson'
import type { PoiCategory } from '../config/poiCategories'
import type { PoiProperties } from './geojson'

export interface QueryCenter {
  lon: number
  lat: number
}

export interface NearbyPoiProperties extends PoiProperties {
  distance_m: number
}

export interface NearbyPoiCollection extends FeatureCollection<Point, NearbyPoiProperties> {
  meta: {
    center: [number, number]
    radius_m: number
    count: number
  }
}

export interface SpatialSummary {
  center: QueryCenter
  radius_m: number
  pois: {
    total: number
    by_category: Partial<Record<PoiCategory, number>>
  }
  buildings: {
    count: number
    footprint_area_m2: number
  }
}

export type SpatialQueryStatus = 'idle' | 'loading' | 'success' | 'empty' | 'error'

export interface FocusPoint extends QueryCenter {
  requestId: number
}

File: backend/app/schemas/spatial.py
from typing import Literal

from pydantic import BaseModel, Field

from app.schemas.geojson import GeoJSONFeature

PoiCategory = Literal[
    "healthcare",
    "emergency",
    "education",
    "public_safety",
    "recreation",
    "transport",
    "commercial",
]
PoiSubcategory = Literal[
    "hospital",
    "clinic",
    "doctors",
    "fire_station",
    "school",
    "college",
    "university",
    "kindergarten",
    "police",
    "park",
    "bus_stop",
    "station",
    "subway_entrance",
    "platform",
    "stop_position",
    "supermarket",
    "mall",
    "marketplace",
]


class NearbyMeta(BaseModel):
    center: tuple[float, float]
    radius_m: int
    count: int


class NearbyPoiCollection(BaseModel):
    type: Literal["FeatureCollection"] = "FeatureCollection"
    features: list[GeoJSONFeature] = Field(default_factory=list)
    meta: NearbyMeta


class QueryCenter(BaseModel):
    lon: float
    lat: float


class PoiSummary(BaseModel):
    total: int
    by_category: dict[str, int]


class BuildingSummary(BaseModel):
    count: int
    footprint_area_m2: float


class SpatialSummary(BaseModel):
    center: QueryCenter
    radius_m: int
    pois: PoiSummary
    buildings: BuildingSummary
